File: mainActivity/src/main/java/com/sierrawireless/avphone/model/AvPhoneObjectData.java
package com.sierrawireless.avphone.model;

import android.text.TextUtils;


public class AvPhoneObjectData {
     public enum Mode {
        None,
        UP,
        DOWN
    }
    private static final String TAG = "AvPhoneModelData";

    public String name;
    public String unit;
    public String defaults;
    public Mode mode;
    public Integer current=null;
    public String label;

    public AvPhoneObjectData(String name, String unit, String defaults, Mode mode, String label) {
        this.name = name;
        this.unit = unit;
        this.defaults = defaults;
        if (isInteger()) {
            this.mode = mode;
            current = Integer.parseInt(defaults);
        }else{
            this.mode = Mode.None;
        }
        this.label = label;
    }

    public String toString() {
        String returned = "{";
        returned = returned + "\"name\": \"" + name + "\",";
        returned = returned + "\"unit\": \"" + unit + "\",";
        returned = returned + "\"defaults\": \"" + defaults + "\",";
        returned = returned + "\"mode\": \"" + modeToString(mode) + "\",";
        returned = returned + "\"label\": \"" + label + "\"}";
        return returned;
    }



    public Boolean isInteger() {
        if (defaults.isEmpty()) { return false;}
        return TextUtils.isDigitsOnly(defaults);
    }

    public String execMode()  {
        if (this.mode == Mode.UP) {
            return increment();
        }else if (this.mode == Mode.DOWN){
            return decrement();
        }
        return defaults;
    }

    public String modeToString(Mode mode){
        switch (mode) {
            case UP:
                return "Increase indefinitely";
            case DOWN:
                return "Decrease to zero";
            case None:
                return "None";
        }
        return "None";
    }

    private String increment(){
        if (current == null) {
            current = Integer.parseInt(defaults);
        }
        current ++;
        return current.toString();
    }

    private String decrement() {
        if (current == null) {
            current = Integer.parseInt(defaults);
        }
        if (current > 0 )
            current --;
        return current.toString();
    }

    public int modePosition() {
        switch (mode) {
            case None:
                return 0;
            case UP:
                return 1;
            case DOWN:
                return 2;
        }
        return 0;
    }

    static public Mode modeFromPosition(int position) {
        switch (position) {
            case 0:
                return Mode.None;
            case 1:
                return Mode.UP;
            case 2:
                return Mode.DOWN;
        }
        return Mode.None;
    }



}

File: mainActivity/src/main/java/com/sierrawireless/avphone/model/AvPhoneObjectData.kt
package com.sierrawireless.avphone.model

import android.text.TextUtils
import com.sierrawireless.avphone.tools.Tools


class AvPhoneObjectData(var name: String, var unit: String, var defaults: String, mode: Mode, private val label: String) {
    var mode: Mode = Mode.None
    var current: Int? = null
    private var min: Int? = null
    private var max: Int? = null


    val isInteger: Boolean
        get() = !defaults.isEmpty() && TextUtils.isDigitsOnly(defaults)

    enum class Mode {
        None,
        UP,
        DOWN,
        RANDOM
    }

    init {
        if (isInteger) {
            this.mode = mode
            current = Integer.parseInt(defaults)
        } else {
            if (mode == Mode.RANDOM) {
                val value = defaults.split(",")
                if (value.size == 2) {
                    if (!value[0].isEmpty() && TextUtils.isDigitsOnly(value[0]) &&
                            !value[1].isEmpty() && TextUtils.isDigitsOnly(value[1])) {
                        min = Integer.parseInt(value[0])
                        max = Integer.parseInt(value[1])
                        this.mode = mode
                        current = Tools.rand(min!!, max!!).toInt()
                    } else {
                        this.mode = Mode.None
                    }
                } else {
                    this.mode = Mode.None
                }
            }else{
                this.mode = Mode.None
            }
        }
    }

    override fun toString(): String {
        var returned = "{"
        returned = "$returned\"name\": \"$name\","
        returned = "$returned\"unit\": \"$unit\","
        returned = "$returned\"defaults\": \"$defaults\","
        returned = returned + "\"mode\": \"" + modeToString(mode) + "\","
        returned = "$returned\"label\": \"$label\"}"
        return returned
    }

    fun execMode(): String {
        return when {
            this.mode == Mode.UP -> increment()
            this.mode == Mode.DOWN -> decrement()
            this.mode == Mode.RANDOM -> random()
            else -> defaults
        }
    }

    private fun modeToString(mode: Mode): String {
        return when (mode) {
            AvPhoneObjectData.Mode.UP -> "Increase indefinitely"
            AvPhoneObjectData.Mode.DOWN -> "Decrease to zero"
            AvPhoneObjectData.Mode.None -> "None"
            AvPhoneObjectData.Mode.RANDOM -> "Random"
        }
    }

    private fun increment(): String {
        if (current == null) {
            current = Integer.parseInt(defaults)
        }
        current = current!! + 1
        return current!!.toString()
    }

    private fun decrement(): String {
        if (current == null) {
            current = Integer.parseInt(defaults)
        }
        if (current!! > 0)
            current = current!! - 1
        else if (current == 0) {
            current = Integer.parseInt(defaults)

        }
        return current!!.toString()
    }

    private fun random(): String {
        current =  Tools.rand(min!!, max!!).toInt()
        return current!!.toString()
    }

    fun modePosition(): Int {
        return when (mode) {
            AvPhoneObjectData.Mode.None -> 0
            AvPhoneObjectData.Mode.UP -> 1
            AvPhoneObjectData.Mode.DOWN -> 2
            AvPhoneObjectData.Mode.RANDOM -> 3
        }
    }

    companion object {

        fun modeFromPosition(position: Int): Mode {
            when (position) {
                0 -> return Mode.None
                1 -> return Mode.UP
                2 -> return Mode.DOWN
                3 -> return Mode.RANDOM
            }
            return Mode.None
        }
    }


}
